Add early support for Ammonite scripts

alexarchambault opened the pull request:
~Just checking if the CI is green for now, will add more details once it is. (This probably needs some more clean-up / tweaking too.)~

Fixes https://github.com/scalameta/metals/issues/1766

alexarchambault commented:
(Quick update: this is almost ready, I'm going to un-draft and add a more lengthy description later today.)

alexarchambault commented:
So, first things first, demo ^^

### Import

![import](https://user-images.githubusercontent.com/7063723/77913325-567d4d00-7294-11ea-9657-751d20ff3339.gif)

### Accurate diagnostics

![diagnostics](https://user-images.githubusercontent.com/7063723/77913393-7876cf80-7294-11ea-8fb5-5a466553b0d1.gif)

### Hover

![hover](https://user-images.githubusercontent.com/7063723/77913446-8f1d2680-7294-11ea-98b2-4a026709abf0.gif)

### Code completion

![completion](https://user-images.githubusercontent.com/7063723/77913370-6bf27700-7294-11ea-8537-7c62cb7b97e4.gif)

### Code navigation

![code navigation](https://user-images.githubusercontent.com/7063723/77913346-6006b500-7294-11ea-8749-a8913642d8ab.gif)

(Don't trust the speed here, the GIFs here feel slower than the original videos.)

alexarchambault commented:
## General idea

This PR adds an `Ammonite` class, that handles most Ammonite-related concerns, most notably:
- "importing" Ammonite scripts,
- starting Ammonite with BSP enabled, via [ammonite-runner](https://github.com/alexarchambault/ammonite-runner),
- converting `*.sc` files to actual scala code that can be passed to the PC,
- adjusting positions (from positions in generated scala code to positions in `.sc` script) in some responses (hover, completion).

The way the Ammonite model is mapped to the BSP model is described in more detail in [this comment](https://github.com/lihaoyi/Ammonite/pull/1053#issuecomment-605952752). Basically, each Ammonite script correspond to a build target. An `import $file` in a script corresponds to a dependency towards that script, just like a module dependency in sbt.

A particularity with Ammonite scripts, is that they can't (or shouldn't) be passed as is to scalac or the PC. When dealing with a script like
```
import $ivy.`com.chuusai::shapeless:2.3.3`
import $file.lib, lib.something
case class Foo(n: Int)
shapeless.Generic[Foo]
```
Ammonite first processes its imports, to spot `import $file`s and `import $ivy`s. It then processes its declarations, so that the final file that it passes to scalac looks like
```
package ammonite
package $file

import _root_.ammonite.interp.api.InterpBridge.{value => interp}
import ammonite.$file.lib

object main{
  import $file.$   , lib.something

  case class Foo(n: Int)
  val res0 = shapeless.Generic[Foo]

  def $main() = { scala.Iterator[String]() }
  override def toString = "main"
}
```

During compilation in Ammonite, the semanticdb plugin generates SemanticDBs for that kind of file. https://github.com/lihaoyi/Ammonite/pull/1053, which adds BSP support in Ammonite, also post-processes SemanticDBs, so that we also get SemanticDBs for the `*.sc` files, with all the positions adjusted.

That discrepancy between `*.sc` / `*.scala` files is an important problem that this PR has to deal with. It does that by
- either ensuring metals doesn't try to generate a SemanticDB for `*.sc` files, but rather uses the Ammonite-post-processed one instead (https://github.com/scalameta/metals/pull/1538/commits/a31501aa3f69b9ce0eab5bec7e0d769f7aa8d1d9, https://github.com/scalameta/metals/pull/1538/commits/98c5f475243bfc5d2d480297b5443aef26d4d2b1), or
- either reading the `*.scala` files generated by Ammonite (https://github.com/scalameta/metals/pull/1538/commits/24b305abe3a7a98ccb9ae40fd55b04f305925977), sometimes regenerating it itself by taking bits of the `.scala` file generated by Ammonite (beginning and end mostly), and generating on-the-fly, in memory, a valid `*.scala` file for some new `*.sc` content (to feed the PC typically, https://github.com/scalameta/metals/pull/1538/commits/9209f0080ec3888c4e165aa2a3e9f24af8fd8e77, https://github.com/scalameta/metals/pull/1538/commits/34b7486b1bab7748cc43c5740f3eab5d820a22e4).

After having passed generated `.scala` files to the PC, the positions it returns have to be adjusted to match the `.sc` file (https://github.com/scalameta/metals/pull/1538/commits/940e21b1e8409e062583c658ef2441415cb320e8, https://github.com/scalameta/metals/pull/1538/commits/d9ed7e7764ae5e995f833f2ea492f6cfbc2bb067

Importing and merging the Ammonite build targets into the default / main build tool build targets is another important problem this PR has to deal with.

Importing Ammonite scripts is proposed to users when they open a not-yet-imported Ammonite script (https://github.com/scalameta/metals/pull/1538/commits/2bcd515be22f7b8bea8decde02397718a3394b7a). Importing can also be triggerred manually via the `ammonite-start` command. When importing is triggered, it looks at all opened buffers, and imports all `.sc` files (except `.worksheet.sc` ones). That also imports all their transitive script dependencies (their `import $file`s, and the `import $file`s of their `import $file`s, etc.).

The build targets provided by Ammonite are blended into the ones of the default build tool, thanks to the refactorings of the very first commits of this PR (https://github.com/scalameta/metals/pull/1538/commits/82c001ccf5f3a821d9c6eb7c1bcf39ab5ce14fc0, https://github.com/scalameta/metals/pull/1538/commits/b4949f9ec4cbff4da107a2b074c74064dc15aad1, https://github.com/scalameta/metals/pull/1538/commits/64fcd59bbef1ea0f8326a015a22ee2facf89e033). I **hope** this doesn't break anything.

`Compilations` needs to be aware of which build server connection it should ask for compilation. This is handled by https://github.com/scalameta/metals/pull/1538/commits/1ed76d0b2d05f0c05acbb7e5ba0fe31241263dc9, https://github.com/scalameta/metals/pull/1538/commits/7faf64d86a55fa0c1dd58eb39ff3dc600dc4a22b. The changes of https://github.com/scalameta/metals/pull/1538/commits/1ed76d0b2d05f0c05acbb7e5ba0fe31241263dc9 might be a bit sensitive in particular, I hope it doesn't break anything either.

Lastly, https://github.com/scalameta/metals/pull/1538/commits/5793a02dcef18b58b888a8f4becbd237157b06f1 deals with Ammonite not being available for `2.12.11` (which should be [addressed](https://github.com/lihaoyi/Ammonite/pull/1041) soon), by defaulting to `2.12.10` for Ammonite scripts, and running a "slow" test for `2.12.10` rather than a "unit" one for `2.12.11`. https://github.com/scalameta/metals/pull/1538/commits/9779957fbc51f8af9d7f8754a0041a72f5906464 allows to pull a custom version of Ammonite, which has https://github.com/lihaoyi/Ammonite/pull/1053.

alexarchambault commented:
## Test it locally

To test that locally, open a scala project in vscode, and put
```json
{
  "metals.serverVersion": "0.8.5-amm-7",
  "metals.customRepositories": [
    "central",
    "bintray:alexarchambault/metals-amm"
  ]
}
```
in `.vscode/settings.json`. Then reload the window, and open a `*.sc` file at the root of the project (scripts in sub-directories should work too, they're just less tested…). An "Ammonite script detected" dialog should open, validate with "Import". The statusbar should then display "Importing Ammonite build", then "Indexing", then "Compiling [script-name].sc". Once compilation is done, hover / completion / code navigation should work, and you should have accurate diagnostics.

alexarchambault commented:
## Limitations

Off the top of my head:

Currently, the way compilation is handled from the BSP server in https://github.com/lihaoyi/Ammonite/pull/1053 is quite sub-optimal (it doesn't cache compilations at all, and re-compiles things all the time, even if they don't need too). That can lead to superfluous compilations of Ammonite scripts, and quirks in code navigation (because Ammonite-generated SemanticDBs were wiped…).

Also, I didn't test much having scripts in sub-directories. Hopefully, it's just a matter of adding a few `relativeTo` here or there in Ammonite…

The `build.sc` scripts of mill are not special-cased yet, so they shouldn't work fine yet (mill adds extra imports in particular).

sjrd commented:
Trying to test this locally, I get the following error in `metals.log`, and so VS Code cannot connect to the build server:
```
ERROR Failed to connect with build server, no functionality will work.
coursierapi.error.SimpleResolutionError$1: Error downloading org.scalameta:mtags_2.13.1:0.8.4-SNAPSHOT
  not found: https://repo1.maven.org/maven2/org/scalameta/mtags_2.13.1/0.8.4-SNAPSHOT/mtags_2.13.1-0.8.4-SNAPSHOT.pom
  not found: https://dl.bintray.com/alexarchambault/metals-amm/org/scalameta/mtags_2.13.1/0.8.4-SNAPSHOT/mtags_2.13.1-0.8.4-SNAPSHOT.pom
  not found: C:\...\.ivy2\local\org.scalameta\mtags_2.13.1\0.8.4-SNAPSHOT\ivys\ivy.xml
  not found: https://oss.sonatype.org/content/repositories/public/org/scalameta/mtags_2.13.1/0.8.4-SNAPSHOT/mtags_2.13.1-0.8.4-SNAPSHOT.pom
  not found: https://oss.sonatype.org/content/repositories/snapshots/org/scalameta/mtags_2.13.1/0.8.4-SNAPSHOT/mtags_2.13.1-0.8.4-SNAPSHOT.pom
```

gabro commented:
@sjrd you'll need to `sbt +publishLocal` for that to work

tgodzik commented:
> Trying to test this locally, I get the following error in `metals.log`, and so VS Code cannot connect to the build server:
> 
> ```
> ERROR Failed to connect with build server, no functionality will work.
> coursierapi.error.SimpleResolutionError$1: Error downloading org.scalameta:mtags_2.13.1:0.8.4-SNAPSHOT
>   not found: https://repo1.maven.org/maven2/org/scalameta/mtags_2.13.1/0.8.4-SNAPSHOT/mtags_2.13.1-0.8.4-SNAPSHOT.pom
>   not found: https://dl.bintray.com/alexarchambault/metals-amm/org/scalameta/mtags_2.13.1/0.8.4-SNAPSHOT/mtags_2.13.1-0.8.4-SNAPSHOT.pom
>   not found: C:\...\.ivy2\local\org.scalameta\mtags_2.13.1\0.8.4-SNAPSHOT\ivys\ivy.xml
>   not found: https://oss.sonatype.org/content/repositories/public/org/scalameta/mtags_2.13.1/0.8.4-SNAPSHOT/mtags_2.13.1-0.8.4-SNAPSHOT.pom
>   not found: https://oss.sonatype.org/content/repositories/snapshots/org/scalameta/mtags_2.13.1/0.8.4-SNAPSHOT/mtags_2.13.1-0.8.4-SNAPSHOT.pom
> ```

To run anything locally you need to cross publish the dependencies using `sbt +publishLocal`

By default only the 2.12.11 mtags jar is published.

sjrd commented:
According to @alexarchambault, the instructions at https://github.com/scalameta/metals/pull/1538#issuecomment-606020007 should be enough to test locally without building *anything* locally. That said, your comment made me try 2.12.11 as my project's `scalaVersion`, and that helps.

However now I run into the following problem:
```
WARN  Error importing Ammonite script C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc
java.util.concurrent.ExecutionException: Boxed Error
	at scala.concurrent.impl.Promise$.resolver(Promise.scala:87)
	at scala.concurrent.impl.Promise$.scala$concurrent$impl$Promise$$resolveTry(Promise.scala:79)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:284)
	at scala.concurrent.Promise.complete(Promise.scala:53)
	at scala.concurrent.Promise.complete$(Promise.scala:52)
	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:187)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: scala.NotImplementedError: an implementation is missing
	at scala.Predef$.$qmark$qmark$qmark(Predef.scala:288)
	at ammrunner.Command$.$anonfun$javaPath$8(Command.scala:47)
	at scala.Option.getOrElse(Option.scala:189)
	at ammrunner.Command$.javaPath(Command.scala:47)
	at ammrunner.Command$Jvm$.fork(Command.scala:131)
	at ammrunner.Command.runBg(Command.scala:22)
	at scala.meta.internal.metals.Ammonite$.$anonfun$socketConn$1(Ammonite.scala:368)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

WARN  no build target for: C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc
WARN  no build target for: C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc
WARN  no build target for: C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc
WARN  code navigation does not work for the file 'C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc' because it doesn't belong to a build target.
WARN  code navigation does not work for the file 'C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc' because it doesn't belong to a build target.
WARN  code navigation does not work for the file 'C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc' because it doesn't belong to a build target.
WARN  code navigation does not work for the file 'C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc' because it doesn't belong to a build target.
WARN  code navigation does not work for the file 'C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc' because it doesn't belong to a build target.
WARN  code navigation does not work for the file 'C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc' because it doesn't belong to a build target.
WARN  code navigation does not work for the file 'C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc' because it doesn't belong to a build target.
WARN  code navigation does not work for the file 'C:\Users\Sepi\Documents\Projets\scalajs-ir-calc\test.sc' because it doesn't belong to a build target.
```
That said, I was able to get autocompletion to run in a .sc file. Diagnostics don't work at all, though.

alexarchambault commented:
@sjrd That should be fixed with `0.8.5-amm-8` (I updated https://github.com/scalameta/metals/pull/1538#issuecomment-606020007). There were issues in the way I published metals to the bintray repository, and in ammonite-runner on Windows.

@gabro @tgodzik I pushed all public metals modules, built with this PR, to [a bintray repository](https://dl.bintray.com/alexarchambault/metals-amm), and added it in https://github.com/scalameta/metals/pull/1538/commits/2928b9439767b201279f67eef056497f4dff5ccc, it ought to work without publishing locally with https://github.com/scalameta/metals/pull/1538#issuecomment-606020007 (it does for me with an empty `~/.ivy2/local`).

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Compilations.scala:
```diff
@@ -39,58 +46,97 @@ final class Compilations(
+
+    val targetsByBuildServer = targets
+      .flatMap(t => buildServer(t).map(_ -> t).toSeq)
+      .groupBy(_._1)
```
Maybe use full value names. It will be easier to figure out.  Plus, let's try to avoid `_1`, `_2` etc.

alexarchambault replied:
Fixed.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Compilations.scala:
```diff
@@ -39,58 +46,97 @@ final class Compilations(
+        Future
+          .successful(Map.empty[BuildTargetIdentifier, b.CompileResult])
+          .asCancelable
+      case (bs, bsTargets) :: Nil =>
```
`buildServer, buildServerTargets`? `bs` has some quite unfortunate conotations. :sweat_smile:

alexarchambault replied:
Fixed.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Compilers.scala:
```diff
@@ -264,12 +275,33 @@ class Compilers(
+        (input, pos)
+      }
+
+      val ammoniteInputPosOpt =
```
Move to a separate function and make both private?

alexarchambault replied:
Done.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala:
```diff
@@ -229,15 +229,21 @@ class DestinationProvider(
-    val parsed =
+    lazy val parsed =
       mtags.index(symbolDefinition.path.toLanguage, defnRevisedInput)
-    if (parsed.occurrences.isEmpty) {
```
You can just add `|| symbolDefinition.path.isAmmoniteScript` here

alexarchambault replied:
Done now (with conditions the other way around).

tgodzik commented on tests/slow/src/test/scala/tests/feature/Ammonite213Suite.scala:
```diff
@@ -0,0 +1,5 @@
+
+import scala.meta.internal.metals.{BuildInfo => V}
+
+class Ammonite213Suite extends tests.BaseAmmoniteSuite(V.scala213)
```
Honestly, I would just keep them all in the same file.

alexarchambault replied:
Done.

tgodzik commented on mtags/src/main/scala/scala/meta/internal/mtags/OnDemandSymbolIndex.scala:
```diff
@@ -145,6 +146,9 @@ final case class OnDemandSymbolIndex(
         val document = mtags.index(language, input)
         s.TextDocuments(List(document))
+      case "sc" =>
+        val res = semanticdb(file)
```
Not 100% sure, but we use Scalameta parser here to create TextDocument on the fly. Shouldn't it also work with `sc` files, but a proper syntax? Or does Scalameta parser not support `.sc` files? Or are the changes so radical that we would need to have a custom one? 

I am worried there can be a race condition between when the file is indexed and senabticDB files is created.

alexarchambault replied:
Just fixed that.

tgodzik commented on tests/unit/src/main/scala/tests/BaseAmmoniteSuite.scala:
```diff
@@ -0,0 +1,238 @@
+        "main.sc",
+        12,
+        18,
+        ".metals/readonly/io/circe/syntax/package.scala"
```
Maybe we can include those as comments in the file? Similar to how we do a lot of the expected suites like `DefinitionLspSuite`?

alexarchambault replied:
I slightly changed `assertDefinitionAtLocation`, so that it infers the exact line / char itself. That should make the resulting code more maintainable.

I didn't inline that like in `DefinitionLspSuite`, for fear the resulting output would change in the future (there are still some definitions, like `import $file.foo, foo.Something`, that don't work as expected yet).

tgodzik commented on tests/unit/src/main/scala/tests/BaseAmmoniteSuite.scala:
````diff
@@ -0,0 +1,238 @@
+      // via Ammonite-generated Semantic DB and indexing of Ammonite-generated source of $file dependencies
+      _ <- server.assertDefinitionAtLocation("main.sc", 14, 25, "main.sc", 6)
+
+      expectedHoverRes = """```scala
````
Would be nice to do it the same way as in `HoverLspSuite`

tgodzik replied:
This can also be a separate test case I think. It could be in a method `checkHovers`

tgodzik commented on mtags/src/main/scala/scala/meta/internal/metals/Docstrings.scala:
```diff
@@ -50,12 +50,15 @@ class Docstrings(index: GlobalSymbolIndex) {
+      source: AbsolutePath,
+      toIndexSource: AbsolutePath
+  ): Unit = {
+    source.toLanguage match {
```
Why do we need a separate `source` to check if this is Scala? Will `toIndexSource` not be a Scala file?

alexarchambault replied:
Good point, should be fixed now.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1572,7 +1583,7 @@ class MetalsLanguageServer(
         buildTargets.linkSourceFile(target, source)
       }
-      indexSourceFile(source, Some(sourceItem))
+      indexSourceFile(source, Some(sourceItem), targets.asScala.headOption)
```
Maybe we could have this:
```scala
    val toIndexSource = targetOpt
      .flatMap(ammonite.generatedScalaPath(_, source))
      .getOrElse(source)
```
in a separate function and just use here to get the real source item?

alexarchambault replied:
Done.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+    }
+
+    // TODO Look at buffers.open rather than focusedDocument?
+    val commandScriptOpt = doc.orElse(focusedDocument()) match {
```
If someone starts AmmoniteServer, they might want to start them all. Relying on whether it's focused doesn't seem that useful for me.

alexarchambault replied:
`focusedDocument` is only used to read the scala and Ammonite version (users can put a comment like `// scala 2.13.1, ammonite 2.0.4` at the top of a script), and only so when users use the `ammonite-start` command directly (when they click "Import" in the "Ammonite script detected" dialog, the script that trigerred the dialog is used instead).

Apart from that, it's not used.

When we start Ammonite, we pass it all the scripts from `buffers.open`. Then, Ammonite loads those, along with their script dependencies. So what is loaded at the end is all opened scripts plus their transitive script dependencies.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+              // do not merge, remove once lihaoyi/Ammonite#1053 is merged
+              .withAmmoniteVersion("2.0.5-bsp-3")
+          )
+        AmmoniteFetcher(versions)
```
Maybe let's move logic to fetch Ammonite to a separate method.

alexarchambault replied:
Done, it's better this way, yes.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+            )
+            Right((command, path))
+        }
+      case Some(path) =>
```
IF we don't rely on focusedDocument, this will not be needed.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+        val scInput = source.toInputFromBuffers(buffers)
+        val input = scalaPath.toInput
+
+        val updatedContent = input.value
```
Could we document how do we generate the source files for PC?

alexarchambault replied:
Done.

alexarchambault commented:
Just saw your two reviews (seems I had missed @tgodzik's one), I'll answer / take them into account.

ckipp01 commented:
Ahhh, also, I forgot to mention that this is incredible work :)

I'm really excited for this feature as I use ammonite scripts pretty often. Also I tested this with `coc-metals` as well, and it seemed to work just as it does in the vscode extension, so that's great ha.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+package scala.meta.internal.metals
```
Sure!

alexarchambault replied:
Should be fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+import java.nio.charset.Charset
+import java.nio.file.Paths
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.{CompletableFuture, TimeoutException, TimeUnit}
```
ok

alexarchambault replied:
Done.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/BuildServerConnection.scala:
```diff
@@ -191,7 +190,6 @@ object BuildServerConnection {
         case conn @ SocketConnection(name, output, input, cancelables) =>
           val tracePrinter = GlobalTrace.setupTracePrinter("BSP")
           val launcher = new Launcher.Builder[MetalsBuildServer]()
-            .traceMessages(tracePrinter)
```
Oh I removed that thinking it was code I had added myself…

alexarchambault replied:
Should be fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Messages.scala:
```diff
@@ -388,4 +388,20 @@ class Messages(icons: Icons) {
   }
+
+  object ImportAmmoniteScript {
+    val doImport: String = "Import"
```
It's even for me, although making that on-demand is probably safer, yes. Note that importing Ammonite scripts also re-indexes the whole workspace, main project included.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -157,6 +161,22 @@ class MetalsLanguageServer(
       parseTrees ::
       compilations.pauseables
   )
+  private val ammonite = new Ammonite(
```
Yes, good point.

alexarchambault replied:
Fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+
+  def loaded(path: AbsolutePath): Boolean =
+    path.isAmmoniteScript && {
+      val uri = path.toNIO.toUri.toASCIIString
```
Mmmhh… that's a problem.

alexarchambault replied:
> `toASCIIString` - will this not be an issue in workspaces using non-ascii characters?

Special characters get encoded in the resulting string, so it's not a problem.
For example, with the character `é`, you get
```scala
@ val s = Paths.get("é.sc").toAbsolutePath.toUri.toASCIIString // encode
s: String = "file://…/e%CC%81.sc"
@ Paths.get(".").toAbsolutePath.normalize.relativize(Paths.get(new URI(s))) // decode back
res12: Path = é.sc
```

alexarchambault replied:
About the problem with `$minus`, I just pushed a commit fixing that in https://github.com/lihaoyi/Ammonite/pull/1053.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+        }
+
+  def start(): CompletableFuture[Object] =
+    start(None)
```
Why not. It's just a make-binary-compatibility-easier-to-preserve habit.

tgodzik replied:
We don't estimate it will ever be important to have it binary compatible. If needed we will expose an API, but since these packages are internal there are no guarantees for anyone who plans to use them outside of Metals.

alexarchambault replied:
Fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+  def start(): CompletableFuture[Object] =
+    start(None)
+
+  def start(doc: Option[AbsolutePath]): CompletableFuture[Object] = {
```
`Future[Unit]` would be better, indeed (and it works fine).

alexarchambault replied:
Fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+    }
+
+  def adjustCompletionList(list: CompletionList, scalaCode: String): Unit = {
+    val adjustPos = adjustPosition(scalaCode)
```
`adjustPosition` is supposed to work fine, and deterministically so.

tgodzik replied:
Then might be better to keep the adjustPosition. You're right.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Compilations.scala:
```diff
@@ -39,58 +46,97 @@ final class Compilations(
   }
 
-  def compileFiles(paths: Seq[AbsolutePath]): Future[b.CompileResult] = {
+  def compileFiles(paths: Seq[AbsolutePath]): Future[Unit] = {
```
See my comment above.

ckipp01 commented:
So, a general question on the UX of this. I've been playing around with this some more and let's say I have the following project:

```
.
├── scripts
│   ├── util.sc
│   └── domain.sc
├── generate.sc
```

More than likely the user will open this up, and let's say the entrypoint is `generate.sc`. They get prompted that `Ammonite Script detected` and they should import. Ok, cool. Then, they switch to `util.sc` and they get prompted again to import. Then they open `domain.sc` and they are prompted again. Since both `util.sc` and `domain.sc` are included in `generate.sc` I'd only expect to be asked to import once, not 3 times for each file which seems to the current behavior correct? How possible would it be to only get prompted that first time, or is each individual script treated separately?

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala:
```diff
@@ -0,0 +1,390 @@
+    if (isCancelled.compareAndSet(false, true)) {
+      val buildShutdown = buildServer0 match {
+        case Some(build) => build.shutdown()
+        case None => Future.successful(())
```
Good point. Note there are still some `Future.successful(())` here or there in metals. The ones here originally came from copied (then updated) code from elsewhere in metals.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/ImportedBuild.scala:
```diff
@@ -1,15 +1,48 @@
+    val d = new DependencySourcesResult(
+      (dependencySources.getItems.asScala ++ other.dependencySources.getItems.asScala).asJava
+    )
+    ImportedBuild(w, o, s, d)
```
Done.

alexarchambault commented:
> Then, they switch to `util.sc` and they get prompted again to import. Then they open `domain.sc` and they are prompted again.

They shouldn't be prompted again. See [my comment above](https://github.com/scalameta/metals/pull/1538#discussion_r403765536). As these are dependencies of a script that was opened when Ammonite was started, they will be loaded, and users shouldn't be prompted, as they'll appear in the build target list once the Ammonite targets are imported.

ckipp01 commented:
> > Then, they switch to `util.sc` and they get prompted again to import. Then they open `domain.sc` and they are prompted again.
> 
> They shouldn't be prompted again. See [my comment above](https://github.com/scalameta/metals/pull/1538#discussion_r403765536). As these are dependencies of a script that was opened when Ammonite was started, they will be loaded, and users shouldn't be prompted, as they'll appear in the build target list once the Ammonite targets are imported.

Alright, clear. It was my mistake. I was being prompted again because the original script had the dash in the name.  Once that was removed it was fine.

alexarchambault commented:
FYI I pushed fixes for most review points. I think the remaining points to discuss are:
- changes in `Compilations`: see my comment [here](https://github.com/scalameta/metals/pull/1538#discussion_r402580647)
- use of `focusedDocument`: it's not used that much, see my comment [here](https://github.com/scalameta/metals/pull/1538#discussion_r403765536)
- should we auto-import instead of opening a dialog: see the discussion below [this comment](https://github.com/scalameta/metals/pull/1538#discussion_r401048628)

ckipp01 commented:
One other thing to maybe address. You mentioned this in the current limitations

> mill: its build.sc scripts are not special-cased yet, so they shouldn't work fine yet (mill adds extra imports in particular).

Using this today, this will no longer detect a mill workspace, but will instead detect an ammonite workspace. So if a mill user is using this, and they dismiss the ammonite prompt, they will no longer be prompted with the "Mill build detected". Any thoughts on specializing at least enough to know that it's a mill space and to prompt the user to import if needed?

alexarchambault commented:
Maybe we can use `BuildTools` to detect if mill is used, and assume `build.sc` is not an Ammonite script in that case.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1708,7 +1729,10 @@ class MetalsLanguageServer(
   }
 
-  private def indexWorkspace(i: ImportedBuild): Unit = {
+  private var lastImportedBuild = ImportedBuild.empty
```
Could we reindex only the Ammonite targets when needed?

alexarchambault replied:
I don't know… It seems that would require being able to remove particular build targets from the indices (those that changed), before adding them back with updated parameters. I'm not sure if that's do-able.

tgodzik commented on mtags/src/main/scala/scala/meta/internal/metals/Docstrings.scala:
```diff
@@ -50,12 +50,14 @@ class Docstrings(index: GlobalSymbolIndex) {
-   * @param path the absolute path for the source file to update.
+   * @param source the absolute path for the source file to update.
    */
-  def expireSymbolDefinition(path: AbsolutePath): Unit = {
```
These changes are not needed now.

alexarchambault replied:
Indeed, should be fixed.

tgodzik commented on tests/unit/src/main/scala/tests/BaseAmmoniteSuite.scala:
```diff
@@ -0,0 +1,287 @@
+      // via Ammonite-generated Semantic DB
+      _ <- server.assertDefinitionAtLocation(
+        "main.sc",
+        "foo.asJson.no|Spaces",
```
Let's use `@@` instead of `|`. That's how we mark the cursor position in the workspace.

alexarchambault replied:
Just changed that.

tgodzik commented on tests/unit/src/main/scala/tests/TestingServer.scala:
```diff
@@ -1281,6 +1281,84 @@ final class TestingServer(
     Assertions.assertNoDiff(obtained, expected)
   }
 
+  def assertDefinitionAtLocation(
```
If these methods are very specific to the Ammonite suite I would move them there.  We already have things like `assertHover` here, so I would like to avoid duplicating the ways of asserting specific things.

We do not have assert definition in TestingServer, but I would go a different way showing the whole file in the test case like shown in the other comment.

alexarchambault replied:
Just moved that to `BaseAmmoniteSuite`.

tgodzik commented on tests/unit/src/main/scala/tests/BaseAmmoniteSuite.scala:
```diff
@@ -0,0 +1,287 @@
+      _ <- server.executeCommand("ammonite-start")
+
+      // via Ammonite-generated Semantic DB
+      _ <- server.assertDefinitionAtLocation(
```
I would rather do something like
```scala
checkDefinition(
s"""|/main.sc
    |$circeImports
    |val foo: Foo = Qux(13, Some(14.0))
    |
    |val json = foo.asJson.no@@Spaces
    |
    |val decodedFoo = decode[Foo](json)
""",
externalDefinition = Some(".metals/readonly/io/circe/syntax/package.scala")
)
```

and if we have a definition in any of local files I would do:

```scala
checkDefinition(
s"""|/main.sc
    |$circeImports
    |val <<foo>>: Foo = Qux(13, Some(14.0))
    |
    |val json = fo@@o.asJson.noSpaces 
    |
    |val decodedFoo = decode[Foo](json)
"""
)
```
We could most likely reuse some existing method for that. Especially for the second case.

I don't mind having separate case classes for each situation.

alexarchambault replied:
I'm not sure about that. Sometimes we go-to-definition from `.metals/readonly` files, that wouldn't work with this way of doing it. Also, if we pass files at that point, we would need to import them, so stop / start Ammonite, which I fear would significantly slow things.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala:
```diff
@@ -0,0 +1,435 @@
+    val versions = versionsOpt
+      .orElse(lastImportVersions)
+      .getOrElse(
+        // TODO Warn if any of these is used
```
TODO? Is it just leftover?

alexarchambault replied:
Yes, just removed it.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala:
```diff
@@ -0,0 +1,435 @@
+        // remove from cancellable too?
+        buildServer0 = None
+        lastImportedBuild0 = ImportedBuild.empty
+        cancelables.remove(value)
```
Maybe we should just do `cancelables.cancel()` this will also remove the value.

alexarchambault replied:
Yes, fixed.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala:
```diff
@@ -0,0 +1,435 @@
+            case Messages.ImportAmmoniteScript.doImport =>
+              start(Some(path)).onComplete {
+                case Failure(e) =>
+                  scribe.warn(s"Error importing Ammonite script $path", e)
```
We should most likely do showMessage too.

alexarchambault replied:
Done.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala:
```diff
@@ -0,0 +1,435 @@
+          else
+            "Current document is not an Ammonite script"
+        scribe.error(msg)
+        Left(new Exception(msg))
```
Can we just do `Future.failed` and then map over it?

alexarchambault replied:
Done.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala:
```diff
@@ -0,0 +1,435 @@
+      case Success(()) =>
+    }
+
+    // TODO Look at buffers.open rather than focusedDocument?
```
Focused document would be when we run the command by hand? If so I think it's fine to have it this way.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala:
```diff
@@ -0,0 +1,435 @@
+        Future.failed(e)
+      case Right((command, script)) =>
+        val extraScripts = buffers.open.toVector
+          .filter(_.isAmmoniteScript)
```
Join the filters?

alexarchambault replied:
Fixed.

tgodzik commented:
Actually, I think I will leave the tests up to you @alexarchambault - we can rework them later on if they prove unwieldy.

alexarchambault commented on build.sbt:
```diff
@@ -483,6 +485,8 @@ def publishMtags =
     .dependsOn(
       crossPublishLocal(V.scala211)
         .dependsOn(crossPublishLocal(V.scala213))
+        // temporary, remove once Ammonite is available for 2.12.11
```
As soon as the release jobs of Ammonite run fine. See the discussion [here](https://github.com/lihaoyi/Ammonite/pull/1057).

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Compilations.scala:
```diff
@@ -38,59 +45,110 @@ final class Compilations(
-  ): Future[b.CompileResult] = {
-    compileBatch(targets)
+  ): Future[Unit] = {
+    compileBatch(targets).map(_ => ())
```
right

alexarchambault replied:
Should be fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1589,17 +1602,28 @@ class MetalsLanguageServer(
   ): Unit = {
+
     try {
+      val toIndexSource0 = toIndexSource(source, targetOpt)
```
I'm not sure it can be moved to `indexWorkspaceSources`, as `indexSourceFile` is called by both `indexWorkspaceSources` and `reindexWorkspaceSources`.

alexarchambault replied:
I changed `toIndexSource` to `sourceToIndex`, it's clearer this way, yes.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Compilations.scala:
```diff
@@ -38,59 +45,110 @@ final class Compilations(
+      case (buildServer, targets) :: Nil =>
+        compile(buildServer, targets)
+          .map(res => targets.map(target => target -> res).toMap)
+      case l =>
```
Fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/Compilations.scala:
```diff
@@ -38,59 +45,110 @@ final class Compilations(
+  ): CancelableFuture[Map[BuildTargetIdentifier, b.CompileResult]] = {
+
+    val targetsByBuildServer = targets
+      .flatMap(t => buildServer(t).map(_ -> t).toSeq)
```
Fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:
```diff
@@ -413,12 +414,83 @@ object MetalsEnrichments
+        case Nil => -1
+        case List(idx) => idx
+        case indices =>
+          throw new Exception(
```
It's better this way, yes. Should be fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:
```diff
@@ -413,12 +414,83 @@ object MetalsEnrichments
+        }
+      }
+
+    def lineAtIndex(index: Int): Int = {
```
Done.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:
```diff
@@ -413,12 +414,83 @@ object MetalsEnrichments
         Paths.get(URI.create(value.stripPrefix("metals:")))
       ).dealias
+
+    def indexToLineChar(index: Int): (Int, Int) = {
```
Indeed, done.

tgodzik commented:
@alexarchambault There are some new conflicts caused by merging the Scala 3 PR. Sorry about that!

Fortunately there are not that many. If you need help with rebasing, let me know. Also, I think we can now concentrate on getting this merged, so this will be my priority in case of reviews.

alexarchambault commented:
@tgodzik No problem, I was actually in the midst of rebasing this, it's going fine.

ckipp01 commented:
So a couple more comment just on usability that I'm curious to get your thoughts on. For reference, I'm using [this](https://github.com/ckipp01/chronica) to test. `webiste.sc` is in the root, and I wanted to see if I could navigate around etc. A couple things I noticed. Some you have referenced above and I'll do my best to hit on them.

- Nested files. You mentioned that nested files should work, but haven't been tested as well. It seems that you can't goto-definition if that definition is in another script? I'm getting the below when I try after a fresh import.

```
2020.04.16 21:32:18 ERROR code navigation does not work for the file '/Users/ckipp/Documents/scala-workspace/chronica/website.sc' because the SemanticDB compiler plugin is not enabled for the build target website.sc.
```
The same for me happens on a hover request for a method defined in another file.

- Bundled libraries. For example, in one my scripts I'm using `ujson`, which I have included without an import. I'm seeing diagnostics that the value isn't found. Are the amm bundled libraries being included?

ckipp01 commented:
One other things I just ran into with the same repo I had listed above. After importing the build and navigating to my `utils.sc` file I tried to do a go-to definition. `coc.nvim` yelled at my due to a uri issue, and when I looked at the trace, it looks like the request that metals is sending back for the location of the definition to jump to is in the actual `.amonnite` directory rather than in the working directory. 

```json
[Trace - 10:45:53 PM] Sending response 'textDocument/definition - (28)'. Processing request took 21ms
Result: [
  {
    "uri": "/Users/ckipp/Documents/scala-workspace/chronica/.ammonite/scala-2.12.10/amm-2.0.5-bsp-4/scripts/utils/src/ammonite/$file/scripts/utils.scala",
    "range": {
      "start": {
        "line": 42,
        "character": 0
      },
      "end": {
        "line": 47,
        "character": 1
      }
    }
  }
]
```

alexarchambault commented:
>  It seems that you can't goto-definition if that definition is in another script? I'm getting the below when I try after a fresh import.

> it looks like the request that metals is sending back for the location of the definition to jump to is in the actual .amonnite directory rather than in the working directory.

Having a look.

> Bundled libraries. For example, in one my scripts I'm using ujson, which I have included without an import. I'm seeing diagnostics that the value isn't found. Are the amm bundled libraries being included?

Ammonite is run as if it was passed `--thin` from here. I think I ran into some issues when running with all its classpath, but I can't remember which right now…

ckipp01 commented:
So one other that I keep meaning to mention here. Is it necessary to prompt the user to import the script upon every open? I feel like this breaks with the current flow of metals where if I'm in a project and my build definition doesn't change at all, then if I close and reopen, I'm not prompted to import again. Only when the build definition changes. I notice that with a script, even if nothing changes, you're prompted to import the build every time you open it.

alexarchambault commented:
@ckipp01 Your [chronica](https://github.com/ckipp01/chronica) repository compiles and works fine for me when:
- adding `import $ivy.``com.lihaoyi::ujson:1.1.0`` ` in `domain.sc`, and
- adding `import ammonite.main.Router._` in `website.sc`

Manually adding `ujson` is required for now, as the BSP server runs Ammonite as if it was passed `--thin`, which isolates the core of Ammonite from what users are seeing. I can't recall the exact problem I ran into when not doing like `--thin` does, maybe it actually works without it now. I'd like to allow users to specify when --thin should be assumed or not, but it's not done yet.

Adding `import ammonite.main.Router._` is required as the BSP server in Ammonite doesn't add the extra imports Ammonite usually adds. That needs to be fixed. I didn't add them initially, as the way Ammonite adds those currently (with some "predef" code) would slow compilation in the BSP server even more, requiring more compilation steps when compiling each script. They could easily be added back by tweaking the way Ammonite handles those (hardcoding the actual imports via a manually defined `ammonite.util.Imports`, rather than going through a predef than needs to be compiled).

alexarchambault commented:
I couldn't reproduce the issue you're seeing with go-to-definition going to a scala file under `.ammonite`. Go-to-definitions seem to work fine for me on your repository (both go-to-usage and go-to-definition, and inside each script or across scripts).

alexarchambault commented:
I guess the first error you were seeing (`code navigation does not work for the file`) must originate from the (some of the) scripts not compiling?

alexarchambault commented:
> So one other that I keep meaning to mention here. Is it necessary to prompt the user to import the script upon every open?

This shouldn't happen, and definitely doesn't from vscode for me. Do you also get that from vscode on your machine?

Unless… When scripts don't parse IIRC, Ammonite won't return any of the scripts as build targets. In that case, the scripts won't be metals' build target list, metals thinks the script is a new one every time, and users are prompted again and again because of that… Ammonite needs to return build targets for the scripts it's passed no matter what. I'll see if I can fix that.

ckipp01 commented:
> adding import $ivy.`com.lihaoyi::ujson:1.1.0` in domain.sc, and
> adding import ammonite.main.Router._ in website.sc

So these fixed almost all of the issues I was having 👍 . I guess my first question then is how will a user debug this issue and know how to add specifically the `ammonite.main.Router._`? I'm not sure I could have figured that out.

Regarding the go-to definition leading you to `.ammonite`, I can only get it every once in a while now with coc-metals when I'm stupidly just navigating like crazy all over the place. I can't get it to happen in vs code.

When trying it with vscode though, one thing I did notice was continual compilation happening or at least `metals/status` is reporting that. I took a video. Keep an eye on the bottom right where the `metals/status` info is shown.

[Amm video](https://i.imgur.com/pyQ7d6V.mp4)

alexarchambault commented:
> When trying it with vscode though, one thing I did notice was continual compilation happening or at least metals/status is reporting that.

This what I meant by "slow compilation, superfluous compilations" in https://github.com/scalameta/metals/pull/1538#issuecomment-606022572 😬 The BSP support in Ammonite, added in https://github.com/lihaoyi/Ammonite/pull/1053, relies on a new script processor added along the way, which doesn't cache compilations at all. It's probably not too hard to address, I guess keeping a checksum of the content of compiled files, and not re-compiling if it didn't change, should be enough. I'll see if I can add that.

alexarchambault commented:
Quick update: I'm waiting for the latest master of Ammonite to be [built and released](https://travis-ci.org/github/lihaoyi/Ammonite/builds/678197666) (seems Travis CI is stuck right now :| ). I'll then use the corresponding Ammonite version here. I hope this will address the transient `DirectoryNotEmptyException` [we're seeing on the Windows job](https://github.com/scalameta/metals/pull/1538/checks?check_run_id=599722489#step:6:4191) right now.

AFAIC, there should be nothing in the way of merging this PR after that.

tgodzik commented:
@alexarchambault Thanks for the awesome work! I think we almost have a great working tool here.

I am checking out the functionality again an having some issues though:
- it seems if you go to definition on value in the script it will go to the generate sources - I think we should return an empty result there, this might be the issue @ckipp01 was having
- if you write for example `Files` and import the suggested java.nio.Files, the completion will break the file:
```
Filesimport java.nio.file.Files
Files
```

or be inserted in a random place:

![files](https://user-images.githubusercontent.com/3807253/80006321-02493f80-84c5-11ea-8d7a-39584294d7ea.gif)

- also encountered a very weird situation where the presentation compiler wasn't suggesting anything but overriding `toString`:

![files](https://user-images.githubusercontent.com/3807253/80006423-23119500-84c5-11ea-8e9f-742fda483800.gif)

- in case values imported from another file, the cursor will not end on the actual symbol:
```
/Basic.sc
val basicValue =@@ 31337
/Other.sc
import $file.Basic
val fileImportVal = Basic.basicValue + 1
```
where `@@` is the cursor postion after going to definition from `Other.sc`

- it seems import are not working unless you run import again. Otherwise, it's all marked as Any. Maybe we should suggest reimport on each `import` addition


Aside from the issues, I think we should add `ammonite script` option to the NewFileProvider, very similar to worksheet one.

tgodzik commented:
@alexarchambault Maybe we would be able to embed ammonite runner inside Metals? For example via java interfaces? I think most issues here is caused by actually using BSP and scripts seem not to benefit from it too much.

ckipp01 commented:
One other thing that I'm noticing and able to reproduce with a very minimal example is that I can continually get hover to fail. For example, lets take the following structure.

```
scritps/
 |- test.sc
main.sc
```
main.sc
```scala
import $file.scripts.test, test.hoverTest

val ho@@ver = hoverTest()
```
test.sc
import $ivy.`com.lihaoyi::scalatags:0.9.0`
import scalatags.Text.all._

def hoverTest = li(a(href := "https://google.com"))
```
With a fresh import, if I execute a hover on `hover`, I'll get nothing.

```[Trace - 10:53:48 PM] Received request 'textDocument/hover - (85)'
Params: {
  "textDocument": {
    "uri": "file:///Users/ckipp/Documents/scala-workspace/amm-test/main.sc"
  },
  "position": {
    "line": 2,
    "character": 6
  }
}


[Trace - 10:53:48 PM] Sending response 'textDocument/hover - (85)'. Processing request took 47ms
Result: null
```

alexarchambault commented:
> Maybe we would be able to embed ammonite runner inside Metals? For example via java interfaces? I think most issues here is caused by actually using BSP and scripts seem not to benefit from it too much.

I'm not sure what the point would be. ammonite-runner does a few non trivial things, like
- parsing scripts to spot comments like `// scala 2.12.10, ammonite 2.0.4`. That allows users to tweak the Ammonite or scala version right from their scripts.
- generating a coursier bootstrap for the right version of Ammonite and scala on-the-fly, rather than doing a simple `java -cp …`. That runs Ammonite as if we pass it `--thin`, which hides the core modules and dependencies of Ammonite from users (`--thin` wouldn't work as is from metals, because Ammonite tweaks the scalac instance beforehand for it). For example, this hides guava, that Ammonite now depends on via bsp4j.

I'm also planning on pushing more for its CLI, which is basically a more powerful `amm` command (able to fetch any Ammonite / scala version, and allowing for more reproducibility via the comments like `// Ammonite 2.0.4, scala 2.12.10`).

alexarchambault commented:
About the issues you spotted, @tgodzik and @ckipp01, do they hinder merging this PR? They're all fix-able. Would you be fine tracking them from issues in the metals repository?

tgodzik commented:
> About the issues you spotted, or the enhancements you propose, @tgodzik and @ckipp01, do they hinder merging this PR? They're all fix-able. Do you think it's fine tracking that kind of issues in the metals repository?

Honestly, I worry about the first impressions people will get if they try out the ammonite support. It would be much nicer to be able to just write and not be much bothered by reimporting each time. I imagine we could even add completions that would be able to get ivy imports from scaladex that you can use right away. BSP in this case seems to be causing more issues than solving.

I would rather wait with adding this feature than merging it as is. We can take over the work here for sure and improve it from now on. This has been a really awesome job anyway, but since we are not in any hurry here, I would rather work on it more.

tgodzik commented:
@alexarchambault so do you think it's impossible to support this feature without the BSP ? Or would that require much more work?

ckipp01 commented:
> About the issues you spotted, or the enhancements you propose, @tgodzik and @ckipp01, do they hinder merging this PR? They're all fix-able. Do you think it's fine tracking that kind of issues in the metals repository?

_If_ we are going to merge, I'd rather do it after the next stable release, and then track all the small things and start tackling them. I agree with what @tgodzik said, if we merge now, I think the first impression people have if they aren't using it on a very simply 1 file script may not be positive, and I think we should be careful with that. Again, I'm really excited about this, but I want to error on the side of it being more solid when we announce it.

alexarchambault commented:
> so do you think it's impossible to support this feature without the BSP ? Or would that require much more work?

BSP maps well to how scripts relate to each other. A script is a build target, depending on a script via an `import $file` is a dependency towards it (like a dependency between modules in sbt). Each script has its own `import $ivy` dependencies, and transitively pulls the ones of its scripts dependencies. It's quite different from the way worksheet works for example.

> If we are going to merge, I'd rather do it after the next stable release, and then track all the small things and start tackling them.

My idea was more to avoid possibly painful rebases in the future. That feature can still be disabled by default, until it's deemed stable enough.

tgodzik commented:
> BSP maps well to how scripts relate to each other. A script is a build target, depending on a script via an `import $file` is a dependency towards it (like a dependency between modules in sbt). Each script has its own `import $ivy` dependencies, and transitively pulls the ones of its scripts dependencies. It's quite different from the way worksheets work for example.

Could we maybe reimport automatically on $file and $ivy imports then?  
 
> My idea was more to avoid possibly painful rebases in the future. That feature can still be disabled by default, until it's deemed stable enough.

I can help out with that for sure, since most conflicts will likely be caused by me. Not sure about merging it and disabling. We could do it now, but what are other people's opinions?

Before merging or enabling it I would for sure do some things still:
- fix the automatic imports - even disable if it's not fixable
- reimport automatically when `$file` or `$import` is added

The rest we can fix afterward for sure.

tgodzik commented:
After thinking about it for a while, I don't think we should merge until it's working reliably. Some minor bugs are fine, but from overall observations I think there are still some basic issues. Could we maybe agree on a roadmap describing what we still need to do?

From what I've seen that encompasses at least:
- spurious compilation issue
- completion auto imports issues
- reimporting automatically (?) or do we really need that? Could we not just ask everything via BSP? We could just restart the presentation compiler if deps change 
- needed `ammonite.main.Router._` - this might be very non-obvious to users
- hover issue that @ckipp01 reported
anything else?

It's really important to make it all right in the first place and I believe this is really not yet ready.

alexarchambault commented:
I'm looking into these issues right now, along with the auto-import.

tgodzik commented:
> I'm looking into these issues right now, along with the auto-import.

Cool! I can help out after I finish with the new release next week. If there are any rebase issues, I can take care of them.

alexarchambault commented:
Some updates: I rebased on the latest master, and fixed most of the issues that were raised here, either via changes in this PR or [in Ammonite](https://github.com/lihaoyi/Ammonite/pull/1072).

Upon opening a script, users are now offered 3 choices rather than 2:
- "Import scripts automatically"
- "Import"
- "Dismiss"

If users choose the first one, scripts get imported automatically later on. I think it's a good middle ground between starting Ammonite without users even noticing, and requiring manual import for every script.

Thanks to `buildTarget/didChange` notifications, the build definitions from Ammonite get automatically reloaded now, when users add / remove / change import `$ivy`s and import `$file`s.

BSP support in Ammonite also caches compilations in memory, so scripts aren't unnecessarily re-compiled. This feels much faster. I guess the next step would be to have BSP support in Ammonite cache the compiler instances in memory. I didn't take the time to do it yet.

We also get accurate tasks for compilation (previously, "Compiling foo.sc" meant that either `foo.sc` or one of its dependency was being compiled. We now get individual "Compiling … .sc", as expected.)

We also get the right Ammonite imports out-of-the-box (no more manual `import ammonite.main.Router._`)

I believe https://github.com/scalameta/metals/pull/1538#issuecomment-619234385 is fixed now, as the issues of https://github.com/scalameta/metals/pull/1538#issuecomment-617880225 (except the second one).

I'd like to tackle the second issue of https://github.com/scalameta/metals/pull/1538#issuecomment-617880225 (imports not being added at the right place). It's very likely users will run into it, from freshly created scripts.

tgodzik commented:
@alexarchambault this feels much better now and something that users will for enjoy to use. 

> it seems if you go to definition on value in the script it will go to the generate sources - I think we should return an empty result there

This is still the case sometimes, but it's not a big deal. We can improve it later on.

> I'd like to tackle the second issue of #1538 (comment) (imports not being added at the right place). It's very likely users will run into it, from freshly created scripts.

I agree, we should for sure try to tackle this still before merging. This is most likely connected to the way we calculate auto imports position in AutoImports class.

@ckipp01 Could you take a look again and check how it feels from an advanced user's perspective?

Some improvements I think we could add later on (we can put it into feature request):
- generate semanticDB files for the scripts
- add an ability to run/debug the script inside Metals

tgodzik commented:
Some other minor things:
- we seem to be getting a log:
```
2020.05.12 18:37:43 ERROR URI 'file:///home/tgodzik/Documents/workspaces/metals-sample/Basic.sc' has invalid format. Example: file:///path/to/base/directory?id=projectName
```
not sure where it's coming from and it doesn't seem to cause any issues, but would be best to get rid of it.
- I see that we actually do generate semanticDB files, but found a small issue:

![image](https://user-images.githubusercontent.com/3807253/81721894-e102c000-9480-11ea-9ed7-4543c2be2b76.png)

it seems occurence is badly generated for `basicValue`, how is the semanticDB plugin added here?

alexarchambault commented:
> we seem to be getting a log:
```
2020.05.12 18:37:43 ERROR URI 'file:///home/tgodzik/Documents/workspaces/metals-sample/Basic.sc' has invalid format. Example: file:///path/to/base/directory?id=projectName
```

I've been seeing them too (a lot actually), but didn't take the time to investigate. These URIs are valid build target identifier [per the spec](https://github.com/build-server-protocol/build-server-protocol/blob/c39d7327f2a4e4c789782d40889ca925effe0c32/docs/specification.md#build-target-identifier). The error message seems to originate from [around here](https://github.com/scalacenter/bloop/blob/d458a63c092dc6ad902e48ffb1f6597c1dd05e85/frontend/src/main/scala/bloop/bsp/ProjectUris.scala#L18-L27) in bloop.

alexarchambault commented:
> it seems occurence is badly generated for `basicValue`, how is the semanticDB plugin added here?

How are you seeing those semanticDBs? Code navigation seems to work for it for me now, and the following gives valid positions:
```bash
$ metap .ammonite/scala-2.12.11/amm-2.1.2-15-b0b582a/Basic/target/META-INF/semanticdb/Basic.sc.semanticdb
Basic.sc
--------

Summary:
Schema => SemanticDB v4
Uri => Basic.sc
Text => non-empty
Language => Scala
Symbols => 5 entries
Occurrences => 2 entries

Symbols:
ammonite/$file/Basic. => final object Basic extends AnyRef { +4 decls }
ammonite/$file/Basic.$main(). => method $main(): Iterator[String]
ammonite/$file/Basic.basicValue. => val method basicValue: Int
ammonite/$file/Basic.basicValue2. => val method basicValue2: Int
ammonite/$file/Basic.toString(). => method toString(): String

Occurrences:
[0:4..0:14): basicValue <= ammonite/$file/Basic.basicValue.
[1:4..1:15): basicValue2 <= ammonite/$file/Basic.basicValue2.
```
with the following scripts:
```scala
$ cat Basic.sc
val basicValue = 31337
val basicValue2 = 31337
$ cat Other.sc
import $file.Basic
val fileImportVal = Basic.basicValue + 1
val fileImportVal2 = Basic.basicValue2 + 1
```

alexarchambault commented:
SemanticDBs are generated with [these options](https://github.com/lihaoyi/Ammonite/blob/2be8fff08c4ff446a6f905553a6259308c9bae34/amm/interp/src/main/scala/ammonite/interp/script/ScriptCompiler.scala#L120-L125). This generates the semanticDBs that can be found at paths like `.ammonite/scala-2.12.11/amm-2.1.2-15-b0b582a/Basic/target/META-INF/semanticdb/ammonite/$file/…` for example. These are the semanticDBs for the `*.scala` files generated by Ammonite, that can be found at paths like `.ammonite/scala-2.12.11/amm-2.1.2-15-b0b582a/Basic/src/ammonite/\$file/Basic.scala`.

These semanticDBs are then post-processed by [this method](https://github.com/lihaoyi/Ammonite/blob/d723a938b7041267a5ae257ead7d15414fb0d096/amm/interp/src/main/scala/ammonite/interp/script/SemanticdbProcessor.scala#L12), to get those corresponding at `*.sc` files, that can be found at paths like `.ammonite/scala-2.12.11/amm-2.1.2-15-b0b582a/Basic/target/META-INF/semanticdb/Basic.sc.semanticdb`.

tgodzik commented:
>  The error message seems to originate from around here in bloop.

It looks like Bloop requires the additional query string and we are not providing it: `?id=projectName`, but are we even compiling the `sc` files using Bloop? I though that came from ammonite?

> How are you seeing those semanticDBs? Code navigation seems to work for it for me now, and the following gives valid positions:

I think the possitions are valid, but we are not mapping them properly for the first value:
```scala
object Basic{
/*<script>*/val basicValue = 1
/*<amm>*/val res_1 = /*</amm>*/basicValue

val secondValues = 23
 /*</script>*/ /*<generated>*/
def $main() = { scala.Iterator[String]() }
  override def toString = "Basic"
  /*</generated>*/
}

```
The first value is moved to the right by the `<script>` and it doesn't look like we are accounting for that. If you use document highlight (the highlight you get when a cursor is on a symbol), you can see what Metals sees.

alexarchambault commented:
> It looks like Bloop requires the additional query string and we are not providing it: `?id=projectName`, but are we even compiling the sc files using Bloop? I though that came from ammonite?

Ammonite doesn't call bloop. Not sure what calls it with the `*.sc` files.

> I think the possitions are valid, but we are not mapping them properly for the first value:

https://github.com/lihaoyi/Ammonite/pull/1072/commits/253d3969a865662b4ce26ef4a658abeee362882f fixed that. It definitely does locally for me.

alexarchambault commented:
(Trying to reproduce locally the issues seen on the Linux and Windows CI…)

tgodzik commented:
> (Trying to reproduce locally the issues seen on the Linux and Windows CI…)

The errors look flaky, so I wouldn't bother too much with it.

alexarchambault commented:
> The errors look flaky, so I wouldn't bother too much with it.

Glad I've been caught by something else in the mean time then ^^

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala:
```diff
@@ -0,0 +1,470 @@
+          _ <- indexWorkspace()
+          _ <- Future.sequence[Unit, List](
+            compilations
+              .cascadeCompileFiles(buffers.open.toSeq) ::
```
Should we only compile ammonite scripts here?

alexarchambault replied:
Probably, yes. Note that the whole workspace (Ammonite scripts + main project) is re-indexed, even though we don't re-import the build for the main project, we just re-use what was already imported. I'm not sure if we need to re-compile the main project or not, in that case.

tgodzik replied:
Since, nothing changes for the Bloop connection, no recompilation should be needed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1551,41 +1584,30 @@ class MetalsLanguageServer(
+      }
+      _ <- profiledIndexWorkspace(() => doctor.check(build.name, build.version))
+      _ = checkRunningBloopVersion(build.version)
+      _ <- Future.sequence[Unit, List](
```
Fixed.

alexarchambault commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -2039,12 +2072,44 @@ class MetalsLanguageServer(
+      },
+      toIndexSource = path => {
+        for {
+          target <- buildTargets.sourceBuildTargets(path).headOption
```
Fixed.

alexarchambault commented:
Found what was causing the `ERROR URI 'file:///….sc' has invalid format. Example: file:///path/to/base/directory?id=projectName` messages. `BuildTargetClasses` was asking the main `BuildServerConnection` about the main and test classes of Ammonite build targets. https://github.com/scalameta/metals/pull/1538/commits/9c405ef42e30edfedf0424112b5e91dd6f46a7c0 and https://github.com/scalameta/metals/pull/1538/commits/cfa01cbf9502be38b5a3abef8a2c5fe30cc5bf9f fix that.
